src/perf_counters.cc

```diff
@@ -19,16 +19,132 @@
 #include <vector>
 
 #if defined HAVE_LIBPFM
+#include <unordered_map>
+
 #include "perfmon/pfmlib.h"
 #include "perfmon/pfmlib_perf_event.h"
 #endif
 
 namespace benchmark {
 namespace internal {
 
-constexpr size_t PerfCounterValues::kMaxCounters;
-
 #if defined HAVE_LIBPFM
+
+class SinglePMURegistry {
+ public:
+  ~SinglePMURegistry() = default;
+  SinglePMURegistry(SinglePMURegistry&&) = default;
+  SinglePMURegistry(const SinglePMURegistry&) = delete;
+  SinglePMURegistry& operator=(SinglePMURegistry&&) noexcept;
+  SinglePMURegistry& operator=(const SinglePMURegistry&) = delete;
+
+  SinglePMURegistry(pfm_pmu_t pmu_id)
+      : pmu_id_(pmu_id), available_counters_(0), available_fixed_counters_(0) {
+    pfm_pmu_info_t pmu_info{};
+    const auto pfm_pmu = pfm_get_pmu_info(pmu_id, &pmu_info);
+
+    if (pfm_pmu != PFM_SUCCESS) {
+      GetErrorLogInstance() << "Unknown pmu: " << pmu_id << "\n";
+      return;
+    }
+
+    name_ = pmu_info.name;
+    desc_ = pmu_info.desc;
+    available_counters_ = pmu_info.num_cntrs;
+    available_fixed_counters_ = pmu_info.num_fixed_cntrs;
+
+    BM_VLOG(1) << "PMU: " << pmu_id << " " << name_ << " " << desc_ << "\n";
+    BM_VLOG(1) << "     counters: " << available_counters_
+               << " fixed: " << available_fixed_counters_ << "\n";
+  }
+
+  const char* name() const { return name_; }
+
+  bool AddCounter(int event_id) {
+    pfm_event_info_t info{};
+    const auto pfm_event_info =
+        pfm_get_event_info(event_id, PFM_OS_PERF_EVENT, &info);
+
+    if (pfm_event_info != PFM_SUCCESS) {
+      GetErrorLogInstance() << "Unknown event id: " << event_id << "\n";
+      return false;
+    }
+
+    assert(info.pmu == pmu_id_);
+
+    if (counter_ids_.find(event_id) != counter_ids_.end()) return true;
+
+    assert(std::numeric_limits<int>::max() > counter_ids_.size());
+    if (static_cast<int>(counter_ids_.size()) >= available_counters_ - 1) {
+      GetErrorLogInstance() << "Maximal number of counters for PMU " << name_
+                            << " (" << available_counters_ << ") reached.\n";
+      return false;
+    }
+
+    counter_ids_.emplace(event_id, info.code);
+
+    BM_VLOG(2) << "Registered counter: " << event_id << " (" << info.name
+               << " - " << info.desc << ") in pmu " << name_ << " ("
+               << counter_ids_.size() << "/" << available_counters_ << "\n";
+
+    return true;
+  }
+
+ private:
+  pfm_pmu_t pmu_id_;
```

Comment:
Nit: can you set `pmu_id_` to an invalid value here? Easier to avoid uninitialized values (even if you set it later in the ctor, setting it at decl is easier to maintain)

or could it be const?